jax/interpreters/pxla.py

```diff
@@ -592,13 +607,45 @@ def _xla_unshard(c, replica_groups, x):
                                   start_indices)
     return c.CrossReplicaSum(padded, replica_groups)
 
-# TODO(mattjj): use more ergonimic form of DynamicUpdateSlice instead!
+# TODO(mattjj): use more ergonomic form of DynamicUpdateSlice instead!
 def _xla_shard_start_indices(c, axis_size, ndim):
   idx = c.Rem(c.ReplicaId(), c.Constant(onp.array(axis_size, onp.uint32)))
   zero = onp.zeros(ndim - 1, onp.uint32)
   return c.Concatenate([c.Reshape(idx, None, [1]), c.Constant(zero)], 0)
 
 
+def _device_put_rep_impl(x, devices, backend=None):
```

Comment:
Shouldn't here (and below) devices be a keyword argument, to match the definition of the primitive? This would also help clarify why we do not have to consider the case when devices and backend are abstract, or in the batch rule why devices is not among batch_dims.

Reply:
Well, being a keyword argument or not is something up to the caller; in Python 2 there's no way to force something to be a keyword argument. (That's distinct from the question of whether an argument has a default value.)

Indeed `devices` here is a parameter of the primitive and, by our convention for `bind`, must be passed as a keyword argument by the caller. But I don't think there's anything that needs changing about this function signature. (Compare e.g. to the [`convert_element_type` primitive](https://github.com/google/jax/blob/b62c3449df4e8226a55a3a04d8432872bceff469/jax/lax/lax.py#L1859-L1876): both `new_dtype` and `old_dtype` are primitive parameters and hence must be passed to `bind` as keyword arguments, but the signature of the primitives' rules can't indicate that.)